pkg/metrics/metrics.jsx

```diff
@@ -917,163 +918,141 @@ class MetricsHour extends React.Component {
+    const dialogInitialProxyValue = runningService(s_pmproxy) && (runningService(s_redis) || runningService(s_redis_server));
     const [dialogError, setDialogError] = useState(null);
-    const [dialogLoggerValue, setDialogLoggerValue] = useState(false);
-    const [dialogProxyValue, setDialogProxyValue] = useState(null);
-    const [dialogInitialProxyValue, setDialogInitialProxyValue] = useState(null);
+    const [dialogLoggerValue, setDialogLoggerValue] = useState(runningService(s_pmlogger));
+    const [dialogProxyValue, setDialogProxyValue] = useState(dialogInitialProxyValue);
     const [pending, setPending] = useState(false);
-    const [deferSaveOnInstall, setDeferSaveOnInstall] = useState(false);
 
-    const s_pmlogger = useObject(() => service.proxy("pmlogger.service"), null, []);
-    const s_pmproxy = useObject(() => service.proxy("pmproxy.service"), null, []);
-    // redis.service on Fedora/RHEL, redis-server.service on Debian/Ubuntu with an Alias=redis
-    const s_redis = useObject(() => service.proxy("redis.service"), null, []);
-    const s_redis_server = useObject(() => service.proxy("redis-server.service"), null, []);
-
-    useEvent(superuser, "changed");
-    useEvent(s_pmlogger, "changed");
-    useEvent(s_pmproxy, "changed");
-    useEvent(s_redis, "changed");
-    useEvent(s_redis_server, "changed");
-
-    let real_redis;
-    let redis_name;
-    if (s_redis_server.exists) {
-        real_redis = s_redis_server;
-        redis_name = "redis-server.service";
-    } else {
-        real_redis = s_redis;
-        redis_name = "redis.service";
-    }
-
-    debug("PCPConfig s_pmlogger.state", s_pmlogger.state, "needs logout", needsLogout, "deferSaveOnInstall", deferSaveOnInstall);
-    debug("PCPConfig s_pmproxy state", s_pmproxy.state, "redis exists", s_redis.exists, "state", s_redis.state, "redis-server exists", s_redis_server.exists, "state", s_redis_server.state);
-
-    if (!superuser.allowed)
-        return null;
+    const handleInstall = () => {
+    // when enabling services, install missing packages on demand
+        const missing = [];
+        if (dialogLoggerValue && !s_pmlogger.exists)
+            missing.push("cockpit-pcp");
+        if (dialogProxyValue && !(s_redis.exists || s_redis_server.exists))
+            missing.push("redis");
+        if (missing.length > 0) {
+            debug("PCPConfig: missing packages", JSON.stringify(missing), ", offering install");
+            Dialogs.close();
+            return install_dialog(missing)
+                    .then(() => {
+                        debug("PCPConfig: package installation successful");
+                        if (missing.indexOf("cockpit-pcp") >= 0)
+                            setNeedsLogout(true);
+                        return wait_cond(() => ((!dialogLoggerValue || s_pmlogger.exists) &&
+                                                (!dialogProxyValue || (s_pmproxy.exists && (s_redis.exists || s_redis_server.exists)))),
```

Comment:
Why does that check the dialog{LoggerProxy}Values here? The dialog is closed at this point, and at this point we have already decided that we want to enable the services and install the packages. I.e. this should just wait on the proxies to appear.

Reply:
You are right, dialogProxyValue might be false. So let's just remove `!dialogLoggerValue ||`.

Reply:
Done.